src/dataflow/operators/ros/to_ros_operator.rs

```diff
@@ -0,0 +1,76 @@
+use crate::dataflow::{context::SinkContext, operator::Sink, Data};
+use serde::Deserialize;
+use std::sync::Arc;
+
+/// Takes an input ERDOS stream and publishes to a ROS topic using the provided message conversion
+/// function.
+///
+/// Conversion function should convert a Rust data type and return a ROS type which implements
+/// the [`rosrust::Message`] trait.
+///
+/// # Example
+/// The following example shows how to use a [`ToRosOperator`] with a conversion function which takes
+/// a Rust [`i32`] and converts it to a ROS message with [`rosrust_msg::std_msgs::Int32`] data.
+///
+/// Assume that `source_stream` is an ERDOS stream sending the correct messages.
+///
+/// ```
+/// fn erdos_int_to_ros_int(input: &i32) -> rosrust_msg::std_msgs::Int32 {
+///     rosrust_msg::std_msgs::Int32 { data: input.data }
+/// }
+///
+/// let ros_sink_config = OperatorConfig::new().name("ToRosInt32");
+/// erdos::connect_sink(
+///     move || -> ToRosOperator<i32, rosrust_msg::std_msgs::Int32> {
+///         ToRosOperator::new("int_topic", erdos_int_to_ros_int)
+///     },
+///     || {},
+///     ros_sink_config,
+///     &source_stream,
+/// );
+/// ```
+
+pub struct ToRosOperator<T, U: rosrust::Message>
+where
+    T: Data + for<'a> Deserialize<'a>,
+{
+    publisher: rosrust::Publisher<U>,
+    to_ros_msg: Arc<dyn Fn(&T) -> U + Send + Sync>,
+}
+
+impl<T, U: rosrust::Message> ToRosOperator<T, U>
+where
+    T: Data + for<'a> Deserialize<'a>,
+{
+    pub fn new<F>(topic: &str, to_ros_msg: F) -> Self
+    where
+        F: 'static + Fn(&T) -> U + Send + Sync,
+    {
+        rosrust::init("publisher");
+        Self {
+            publisher: rosrust::publish(topic, 2).unwrap(),
+            to_ros_msg: Arc::new(to_ros_msg),
+        }
+    }
+}
+
+impl<T, U: rosrust::Message> Sink<(), T> for ToRosOperator<T, U>
+where
+    T: Data + for<'a> Deserialize<'a>,
+{
+    fn on_data(&mut self, ctx: &mut SinkContext<()>, data: &T) {
+        let timestamp = ctx.get_timestamp().clone();
+        let msg = (self.to_ros_msg)(data);
+
+        tracing::info!(
+            "{} @ {:?}: Sending {:?}",
+            ctx.get_operator_config().get_name(),
+            timestamp,
+            msg,
+        );
+        // Publishes converted message on topic.
+        self.publisher.send(msg).unwrap();
+    }
+
+    fn on_watermark(&mut self, _ctx: &mut SinkContext<()>) {}
```

Comment:
Similar in spirit to the exposing of timestamps to the `to_ros_msg` function, we should be able to send watermarks downstream. 

Although this might be a more difficult endeavor, does ROS have Enum typed messages?

Reply:
ROS/ROS2 don't support Enum typed messages. [Here's](https://answers.ros.org/question/9427/enum-in-msg/) a workaround using constants.

Reply:
Maybe we can have the`to_ros_msg` function handle watermarks as well, since it already takes an ERDOS message? One potential issue is that each watermark must produce a ROS message, since the function type is `&erdos_msg -> ros_msg`.

I propose we change the function type to `&erdos_message -> Vec<ros_message>`. This way, we have maximum flexibility because both watermarks and data messages could produce 0, 1, or several ERDOS messages.

Likewise, to match the API, I also propose we change the function type for `to_erdos_msg` (which should be renamed to `from_ros_msg`) to `&ros_msg -> Vec<erdos_msg>`. This will support useful patterns like sending data followed by a watermark.

@sukritkalra @objorkman thoughts?

Reply:
As discussed in the ERDOS meeting, we are in agreement on this change. Let's make sure that both `to_ros_msg` and `from_ros_msg` take `Vec<Message<T>>` and that the same method is invoked for both `Message` and `WatermarkMessage` callbacks here.